db/hints/manager.cc

```diff
@@ -223,32 +223,32 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
     co_return;
 }
 
-bool manager::end_point_hints_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) noexcept {
+bool manager::end_point_hints_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state, uint64_t operation_id) noexcept {
```

Comment:
The whole series looks good in general, but I would prefer not to have this `operation_id` trick for two reasons:
 1. it bloats the interface
 2. it piggy-backs on fake stats added just to provide a generator for this id

Perhaps you could treat `fm.get()`, the underlying pointer from the frozen mutation, as operation id? It can of course get reused by the allocator later, but it should be clear from the flow of operations whether they belong to the same hint, or a new one.

Reply:
Yes, a pair of (endpoint address, mutation memory address) should be sufficient to differentiate (the address is needed because the same mutation may be stored for multiple endpoints at the same time).

I'll switch to using mutation addresses.

Reply:
Hmm, it seems that on the hint send path we are not passing the mutation through a pointer, we are moving `frozen_mutation_and_schema` directly. I guess we can change it so that we keep it behind a `std::unique_ptr`. It will be an additional small allocation per hint send operation, but hint sending is asynchronous and not performance critical, so I think it's OK.

Reply:
On the other hand, introducing an allocation in order just to have a stable address for identification purposes seems wrong to me. Removing the `operation_id` only on the store path and not on the send path leaves an ugly asymmetry. I'm not sure what approach would be the best here.

@psarna @vladzcloudius @haaawk thoughts on this?

Reply:
Frozen mutation is generated too late - in send_one_hint(). 
But there is a problem with this patch'es approach in general: you pass around something that is only used for debug/trace logging and this makes very little sense.

Whatever you use for such a logging you don't want this to consume resources when such a logging is not enabled. In your case it would consume resources even if it's not needed.

Let's try to get back to the problem you are trying to resolve: you want to be able to identify log messages that are related to the same object - stored or sent hint. You had decided that you need an ID that would be unique in all relevant scopes and work hard to achieve that - but is it really necessary?

You could simply print a message at a corresponding log level that would "connect" an ID from one scope (e.g. file name and replay position) to ID in a deeper scope (e.g. a frozen mutation address). And then use an ID already available in the corresponding scope to identify corresponding log messages.

This will allow you not to waste resources when not needed and still be able to track messages that relate to a specific hint object with a trivial post-processing of the log.